File: exercises/03 PhotoSearcher/C#/03 PhotoSearcher/Color.cs
namespace _03_PhotoSearcher
{
    public class Color
    {
        public int Red { get; }
        private int Green { get; }
        private int Blue { get; }

        public Color(int red, int green, int blue)
        {
            Red = red;
            Green = green;
            Blue = blue;
        }

        protected bool Equals(Color other)
        {
            return Red == other.Red && Green == other.Green && Blue == other.Blue;
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != this.GetType()) return false;
            return Equals((Color)obj);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                var hashCode = Red;
                hashCode = (hashCode * 397) ^ Green;
                hashCode = (hashCode * 397) ^ Blue;
                return hashCode;
            }
        }
    }
}

File: exercises/03 PhotoSearcher/C++/03 PhotoSearcher/Color.cpp
#include "Color.h"

Color::Color(int red, int green, int blue) :
	_red(red),
	_green(green),
	_blue(blue)
{
}

Color::~Color()
{
}

bool Color::operator==(const Color& other) const
{
	return
		_red == other._red &&
		_green == other._green &&
		_blue == other._blue;
}
